### DISCUSSION TITLE
Controlling MOBIKE behavior

### DISCUSSION THREAD

**Discussion by hgshoggins** [strongswan/strongswan#1625] [Q&A] [answered]
We have a box that acts as an initiator to establish several net-to-net tunnels. This box has a main connection, established on PPPoE (with `ppp0` appearing and disappearing), and a backup one, on a permanent `eth` interface. 

In a change of default route scenario (happens from time to time as the PPPoE sometimes get dropped for a few minutes), there are two cases : one when `ppp0` becomes the default route, and one when it's `eth0`. 

For the latter, everything works like a charm: when the backup connexion is up and `ppp0` disappears, `charon` says (`Y.Y.Y.Y` is the responder):
```
Mar 25 18:31:51 yomama charon: 07[IKE] old path is not available anymore, try to find another
Mar 25 18:31:51 yomama charon: 07[IKE] looking for a route to Y.Y.Y.Y ...
Mar 25 18:31:51 yomama charon: 07[IKE] requesting address change using MOBIKE
Mar 25 18:31:51 yomama charon: 07[ENC] generating INFORMATIONAL request 2 [ ]
```
... and we keep all the tunnels, the change of route is unnoticeable. Yay. 

But when `ppp0` appears it's not exactly working as expected. `charon`'s output is not exactly the same (notice the difference between `requesting address change using MOBIKE` before and now, `sending address list update using MOBIKE`): 
```
Mar 25 18:32:16 yomama charon: 16[KNL] 10.1.1.20 appeared on ppp0
Mar 25 18:32:16 yomama charon: 05[IKE] sending address list update using MOBIKE
Mar 25 18:32:16 yomama charon: 05[ENC] generating INFORMATIONAL request 4 [ N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) ]
```
Okay, `charon` sends a new address list, but then the responder gets a bit angry at us (`X.X.X.X` being the correct new outgoing IP address of the initiator):
```
Mar 25 18:32:31 webfront-2 charon[2329891]: 08[NET] received packet: from X.X.X.X[500] to Y.Y.Y.Y[500] (392 bytes)
Mar 25 18:32:31 webfront-2 charon[2329891]: 08[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
Mar 25 18:32:31 webfront-2 charon[2329891]: 08[IKE] X.X.X.X is initiating an IKE_SA
Mar 25 18:32:31 webfront-2 charon[2329891]: 08[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_AES128_XCBC/CURVE_25519
Mar 25 18:32:31 webfront-2 charon[2329891]: 08[IKE] remote host is behind NAT
Mar 25 18:32:31 webfront-2 charon[2329891]: 08[IKE] sending cert request for "<redacted>"
Mar 25 18:32:31 webfront-2 charon[2329891]: 08[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
Mar 25 18:32:31 webfront-2 charon[2329891]: 08[NET] sending packet: from Y.Y.Y.Y[500] to X.X.X.X[500] (273 bytes)
Mar 25 18:32:32 webfront-2 charon[2329891]: 03[NET] ignoring IKE_SA setup from X.X.X.X, peer too aggressive
Mar 25 18:32:37 webfront-2 charon[2329891]: 03[NET] ignoring IKE_SA setup from X.X.X.X, peer too aggressive
Mar 25 18:32:38 webfront-2 charon[2329891]: 03[NET] ignoring IKE_SA setup from X.X.X.X, peer too aggressive
Mar 25 18:32:40 webfront-2 charon[2329891]: 03[NET] ignoring IKE_SA setup from X.X.X.X, peer too aggressive
Mar 25 18:32:41 webfront-2 charon[2329891]: 03[NET] ignoring IKE_SA setup from X.X.X.X, peer too aggressive
Mar 25 18:32:43 webfront-2 charon[2329891]: 03[NET] ignoring IKE_SA setup from X.X.X.X, peer too aggressive
Mar 25 18:32:44 webfront-2 charon[2329891]: 03[NET] ignoring IKE_SA setup from X.X.X.X, peer too aggressive
Mar 25 18:32:45 webfront-2 charon[2329891]: 03[NET] ignoring IKE_SA setup from X.X.X.X, peer too aggressive
```

So, apparently on the initiator side, MOBIKE behaves differently whether the change of route is due to an interface showing up or disappearing, and/or just replacing the default route with one another. 

What do you suggest I should do to have `charon` request an address change rather than just sending an address list update?

**Comment by tobiasbrunner** [comment, marked as answer]
> Okay, `charon` sends a new address list, but then the responder gets a bit angry at us (`X.X.X.X` being the correct new outgoing IP address of the initiator):

I don't see how these messages are related. The responder's log messages there are about a new IKE_SA being created, not any INFORMATIONAL request. Check if you find anything about that in the initiator's log.

> So, apparently on the initiator side, MOBIKE behaves differently whether the change of route is due to an interface showing up or disappearing, and/or just replacing the default route with one another.

If the current path is still available, strongSwan won't change anything by default. It just notifies the responder about any possible new alternative IP addresses. You can change this behavior by enabling `charon.prefer_best_path` in [strongswan.conf](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html). (Whether it has the desired effect depends on the routes.)

**Reply by hgshoggins** [reply]
As for the first symptom found, on the initiator side, there was a lot of:
```
Mar 25 18:32:20 yomama charon: 06[IKE] retransmit 1 of request with message ID 0
Mar 25 18:32:20 yomama charon: 06[NET] sending packet: from 10.1.1.20[500] to Y.Y.Y.Y[500] (392 bytes)
Mar 25 18:32:21 yomama charon: 07[KNL] creating delete job for CHILD_SA ESP/0x00000000/Y.Y.Y.Y
Mar 25 18:32:21 yomama charon: 07[JOB] CHILD_SA ESP/0x00000000/Y.Y.Y.Y not found for delete
Mar 25 18:32:22 yomama charon: 09[KNL] creating acquire job for policy 192.168.22.10/32[udp/48285] === 192.168.66.10/32[udp/commplex-main] with reqid {1}
Mar 25 18:32:22 yomama charon: 09[CFG] ignoring acquire for reqid 1, connection attempt pending
Mar 25 18:32:22 yomama charon: 11[IKE] giving up after 1 retransmits
Mar 25 18:32:22 yomama charon: 11[IKE] peer not responding, trying again (2/0)
Mar 25 18:32:22 yomama charon: 11[IKE] initiating IKE_SA net-net[1] to Y.Y.Y.Y
Mar 25 18:32:22 yomama charon: 11[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
Mar 25 18:32:22 yomama charon: 11[NET] sending packet: from 10.1.1.20[500] to Y.Y.Y.Y[500] (392 bytes)
```

Anyway the problem is solved with your suggestion: `charon` detects the route changes because they're changed through a script when the PPP link is up and the `charon.prefer_best_path = yes` option was the key. Now I get the expected `requesting address change using MOBIKE`.

**Reply by tobiasbrunner** [reply]
That only shows the IKE_SA_INIT, not why there was one in the first place. Anyway, [one retransmit](https://docs.strongswan.org/docs/5.9/config/retransmission.html) might be a bit optimistic.
